src/sqlfluff/core/rules/analysis/select_crawler.py

```diff
@@ -29,15 +40,40 @@ def gather(
         cls, segment: BaseSegment, dialect: Dialect
     ) -> Dict[Optional[str], List["SelectCrawler"]]:
         """Find top-level SELECTs and CTEs, return info."""
-        queries = defaultdict(list)
+        queries: DefaultDict[Optional[str], List[SelectCrawler]] = defaultdict(list)
+        cls._process_select_statements(segment, dialect, queries)
+        return dict(queries)
+
+    @classmethod
+    def _process_select_statements(
+        cls,
+        segment: BaseSegment,
+        dialect: Dialect,
+        queries: DefaultDict[Optional[str], List["SelectCrawler"]],
+    ):
         # We specify recurse_into=False because we only want top-level select
-        # statmeents and CTEs. We'll deal with nested selects later as needed,
+        # statements and CTEs. We'll deal with nested selects later as needed,
         # when processing their top-level parent.
         for select_statement in segment.recursive_crawl(
             "select_statement", recurse_into=False
         ):
-            select_name = cls._get_name_if_cte(select_statement, segment)
-            queries[select_name].append(SelectCrawler(select_statement, dialect))
+            select_statement_info = cls._get_cte_info(select_statement, segment)
+            if select_statement_info:
+                queries[select_statement_info.select_name].append(
+                    SelectCrawler(select_statement, dialect, select_statement_info.cte)
+                )
+
+    @classmethod
+    def gather_nested_ctes(
```

Comment:
New function, similar to the existing `gather()` function, but this one handles nested CTEs.